runtime/storage.go

```diff
@@ -96,279 +101,142 @@ func NewStorage(ledger atree.Ledger, memoryGauge common.MemoryGauge) *Storage {
-	// If domain exists (domain register exists), domain storage map is loaded and returned.
-	//
-	// If domain doesn't exist and createIfNotExists is true,
-	// new domain storage map is created and returned.
-	// NOTE: given account would be migrated in Commit() since this is write op.
-	//
-	// If domain doesn't exist and createIfNotExists is false, nil is returned.
-	//
-	//
-	// ### Migration of unmigrated accounts
-	//
-	// Migration happens in Commit() if unmigrated account has write ops.
-	// NOTE: Commit() is not called by this function.
-	//
-	// Specifically,
-	// - unmigrated account is migrated in Commit() if there are write ops.
-	//   For example, inserting values in unmigrated account triggers migration in Commit().
-	// - unmigrated account is unchanged if there are only read ops.
-	//   For example, iterating values in unmigrated account doesn't trigger migration,
-	//   and checking if domain exists doesn't trigger migration.
-
 	// Get cached domain storage map if it exists.
 
 	domainStorageKey := interpreter.NewStorageDomainKey(s.memoryGauge, address, domain)
 
-	if domainStorageMap := s.cachedDomainStorageMaps[domainStorageKey]; domainStorageMap != nil {
-		return domainStorageMap
-	}
-
-	// Get (or create) domain storage map from existing account storage map
-	// if account is migrated account.
-
-	accountStorageKey := interpreter.NewStorageKey(s.memoryGauge, address, AccountStorageKey)
-
-	accountStorageMap, err := s.getAccountStorageMap(accountStorageKey)
-	if err != nil {
-		panic(err)
-	}
-
-	if accountStorageMap != nil {
-		// This is migrated account.
-
-		// Get (or create) domain storage map from account storage map.
-		domainStorageMap := accountStorageMap.GetDomain(s.memoryGauge, inter, domain, createIfNotExists)
-
-		// Cache domain storage map
+	if s.cachedDomainStorageMaps != nil {
+		domainStorageMap = s.cachedDomainStorageMaps[domainStorageKey]
 		if domainStorageMap != nil {
-			s.cachedDomainStorageMaps[domainStorageKey] = domainStorageMap
+			return domainStorageMap
 		}
-
-		return domainStorageMap
-	}
-
-	// At this point, account is either new or unmigrated account.
-
-	domainStorageMap, err := getDomainStorageMapFromLegacyDomainRegister(s.Ledger, s.PersistentSlabStorage, address, domain)
-	if err != nil {
-		panic(err)
 	}
 
-	if domainStorageMap != nil {
-		// This is a unmigrated account with given domain register.
-
+	defer func() {
 		// Cache domain storage map
-		s.cachedDomainStorageMaps[domainStorageKey] = domainStorageMap
-
-		// Add account to unmigrated account list
-		s.addUnmigratedAccount(address)
-
-		return domainStorageMap
-	}
-
-	// At this point, account is either new account or unmigrated account without given domain.
+		if domainStorageMap != nil {
+			s.cacheDomainStorageMap(
+				domainStorageKey,
+				domainStorageMap,
+			)
+		}
+	}()
 
-	// Domain doesn't exist.  Return early if createIfNotExists is false.
+	if s.IsV1Account(address) ||
+		!s.Config.StorageFormatV2Enabled {
```

Comment:
Maybe we can:
- use AccountStorageV1 for accounts with domain registers
- use AccountStorageV2 for accounts with "stored" register
- use either AccountStorageV1 or AccountStorageV2 for new accounts depending on `s.Config.StorageFormatV2Enabled`

This will allow us to have consistent account format for existing accounts regardless of `s.Config.StorageFormatV2Enabled` settings.

Reply:
If I understand this correctly, are you proposing to use `AccountStorageV2` for accounts with `"stored"` register, even if the storage format V2 feature is not enabled?

If that's the case, we won't be able to roll out this feature without an HCU. If we want to roll it out without an HCU, we cannot use `AccountStorageV2` if the feature is not enabled. If the feature is not enabled, it must behave exactly like the current code, which only supports V1.

Reply:
> are you proposing to use AccountStorageV2 for accounts with "stored" register, even if the storage format V2 feature is not enabled?

Yes. 

If account is in V2 format already and feature is disabled for any reason, do we read account data in V2 format or treat it as no-data?  Or maybe we don't need to be concerned that feature is disabled after it is enabled?

> If that's the case, we won't be able to roll out this feature without an HCU. If we want to roll it out without an HCU, we cannot use AccountStorageV2 if the feature is not enabled. If the feature is not enabled, it must behave exactly like the current code, which only supports V1.

I think we can still roll out feature (disabled by default) without an HCU because the only difference is how to handle account in V2 when feature is disabled after it was enabled long enough to migrate some accounts.

When rolled out with feature disabled, all accounts are in V1 and will continue to behave as V1.  When feature is enabled, accounts (that satisfy criteria) will migrate to V2.

If feature is disabled after it is enabled, there can be a mix of V1 and V2 accounts.

I.e. If the feature becomes re-disabled, we can handle V1 as V1, handle V2 as V2, but we would stop migrating any existing V1 to V2, and use V1 for new accounts.

Reply:
> If account is in V2 format already and feature is disabled for any reason, do we read account data in V2 format or treat it as no-data? Or maybe we don't need to be concerned that feature is disabled after it is enabled?

That would mean that at some point the some program was executed where the feature was enabled, and now it is turned off again. That should never occur and we should treat that as an implementation error, as we have no means to "migrate back" and I don't think we should support that to start with (we still might need to add that in case something goes wrong, but lets consider that when it's actually the case).

Reply:
> That would mean that at some point the some program was executed where the feature was enabled, and now it is turned off again. That should never occur and we should treat that as an implementation error,

OK.  In that case, maybe we can modify `IsV1Account()` to `AccountStatus()` which returns either v1, v2, or new status, and also confirm account is not v2 when feature is disabled.  This would only read one more register.

Basically, I am just looking out for any situation that might cause the account to be in an inconsistent state.